aws-usage-report/aws-usage-report.qmd

```diff
@@ -0,0 +1,356 @@
+    y = "Size (GiB)"
+  )
+
+```
+
+### Compute costs and usage
+
+```{r ec2-costs}
+# https://www.paws-r-sdk.com/docs/costexplorer_get_cost_and_usage/
+# https://docs.aws.amazon.com/aws-cost-management/latest/APIReference/API_GetDimensionValues.html
+
+# TODO: modify ce_to_df to deal with an arbitrary number of metrics so
+# we can do this in one call with `Metrics = list("UnblendedCost", "UsageQuantity")
+# and pass it to ce_to_df() once, rather than joining
+ec2_instance_type_costs_usage_res <- cost_explorer$get_cost_and_usage(
+  TimePeriod = list(Start = start_date, End = end_date),
+  Granularity = "DAILY",
+  Filter = list(
+    Dimensions = list(
+      Key = "RECORD_TYPE",
+      Values = "Usage"
+    ),
+    Dimensions = list(
+      Key = "SERVICE",
+      Values = "Amazon Elastic Compute Cloud - Compute"
+    )
+  ),
+  Metrics = list("UnblendedCost", "UsageQuantity"),
+  GroupBy = list(
+    list(
+      Type = "DIMENSION",
+      Key = "SERVICE"
+    ),
+    list(
+      Type = "DIMENSION",
+      Key = "INSTANCE_TYPE"
+    )
+  )
+)
+
+# Join costs and usage hours
+ec2_instance_type_costs_usage <- ec2_instance_type_costs_usage_res |>
+  ce_to_df(metric = "UnblendedCost") |>
+  left_join(
+    ce_to_df(ec2_instance_type_costs_usage_res, metric = "UsageQuantity"),
+    by = c("start_date", "end_date", "service", "instance_type")
+  ) |>
+  filter(
+    instance_type != "NoInstanceType"
+  )
+
+instances <- unique(ec2_instance_type_costs_usage$instance_type)
+
+instance_palette <- paletteer::paletteer_d(
+  "ggpomological::pomological_palette",
+  n = length(instances)
+) |>
+  setNames(instances)
+```
+
+```{r daily-usage-by-instance}
+ec2_instance_type_costs_usage |>
+  mutate(instance_type = fct_reorder(instance_type, desc(UsageQuantity), .fun = max, .desc = TRUE)) |>
+  ggplot(aes(x = start_date, y = UsageQuantity, fill = instance_type)) +
+  geom_col() +
+  scale_fill_manual(values = instance_palette) +
+  labs(
+    title = "Daily EC2 usage by instance type",
+    x = "Date",
+    y = "Usage (hours)",
+    fill = "EC2 Instance Type"
+  )
+```
+
+```{r daily-cost-by-instance}
+ec2_instance_type_costs_usage |>
+  mutate(instance_type = fct_reorder(instance_type, desc(UnblendedCost), .fun = max, .desc = TRUE)) |>
+  ggplot(aes(x = start_date, y = UnblendedCost, fill = instance_type)) +
+  geom_col() +
+  scale_fill_manual(values = instance_palette) +
+  labs(
+    title = "Daily EC2 cost by instance type",
+    x = "Date",
+    y = "Daily Cost ($)",
+    fill = "EC2 Instance Type"
+  )
+```
+
+```{r total-usage-vs-cost-by-instance}
+ec2_instance_type_costs_usage |>
+  group_by(instance_type) |>
+  summarize(
+    total_hours = sum(UsageQuantity),
+    total_cost = sum(UnblendedCost)
+  ) |>
+  ggplot(aes(x = total_hours, y = total_cost, colour = instance_type)) +
+  geom_point(size = 5) +
+  scale_colour_manual(values = instance_palette) +
+  labs(
+    title = "Total cost vs hours on different EC2 instance types",
```

Comment:
Looks like title and Y axis don't align, please double-check

Reply:
Good catch, thanks!